src/google/google-extractor/api/api_caller.py

```diff
@@ -0,0 +1,45 @@
+from opnieuw import retry
+from tail_recursive import tail_recursive
+
+
+@retry(
+    retry_on_exceptions=(Exception),
+    max_calls_total=4,
+    retry_window_after_first_call_in_seconds=60,
+)
+def execute(executable_resource):
+    return executable_resource.execute()
+
+
+@tail_recursive
+def call_api(resource_method, resource_parameters, response_property, results=None):
+    """
+    Call a Google Classroom/Admin SDK API
+
+    Parameters
+    ----------
+    resource_method: function
+        is the get/list SDK function to call
+    resource_parameters: dict
+        is the parameters for get/list
+    response_property: string
+        is the property in the API response we want
+    results: list
+        is the list of dicts of the API response accumulated across pages
+
+    Returns
+    -------
+    list
+        a list of dicts of the API response property requested
+    """
+    if results is None:
+        results = []
+    response = execute(resource_method(**resource_parameters))
```

Comment:
How are 4xx and 5xx responses handled?

Reply:
in this code, 4xx and 5xx are blanket retried with exponential backoff with a 60 second max.  then it crashes with an uncaught exception (oops).  will update to instead log the exception and move on.  future code will be finer grained e.g. handle 429 rate limiting, 401 unauthorized, etc.

Reply:
Uncaught exception might be good enough for now. I believe we have ticket for adding logging to the application later.